picard/tagger.py

```diff
@@ -346,105 +363,152 @@ def move_files(self, files, target):
             self.cluster(files)
 
     def add_files(self, filenames, target=None):
+        self._add_files([self._add_files_from_directory("", filenames)], target=target)
+
+    def _add_files(self, list_of_filenames, target=None):
         """Add files to the tagger."""
-        ignoreregex = None
-        pattern = config.setting['ignore_regex']
-        if pattern:
-            ignoreregex = re.compile(pattern)
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        new_files = []
+        number_of_files = sum([len(x) for x in list_of_filenames])
+        number_of_directories = len(list_of_filenames)
+        log.debug("Adding a total of %d files in %d directories",
+                number_of_files, number_of_directories)
+        self.window.set_statusbar_message(
+            N_("Adding a total of %(files)d files in %(directories)d directories ..."),
+            {'files': number_of_files, 'directories': number_of_directories},
+            echo=None
+        )
+        for filenames in list_of_filenames:
+            thread.run_task(partial(self._open_files, filenames),
+                    partial(self._open_files_finished, target),
+                    priority=2,
+                    thread_pool=self.load_thread_pool)
+
+    def _open_files(self, filenames):
+        new_files = {}
         for filename in filenames:
             filename = os.path.normpath(os.path.realpath(filename))
-            if ignore_hidden and is_hidden(filename):
-                log.debug("File ignored (hidden): %r" % (filename))
-                continue
-            if ignoreregex is not None and ignoreregex.search(filename):
-                log.info("File ignored (matching %r): %r" % (pattern, filename))
-                continue
             if filename not in self.files:
                 file = open_file(filename)
                 if file:
-                    self.files[filename] = file
-                    new_files.append(file)
-        if new_files:
+                    new_files[filename] = file
+                # Allow UI to process events to keep it responsive
+                QtCore.QCoreApplication.processEvents()
```

Comment:
I have no idea - I didn't benchmark how often it needed to be done on my system for the UI to stay responsive, but it would depend mainly on the time taken to read the 128 bytes for the file (so on whether you had an SSD or were accessing a file on a share) and to a lesser extent on your CPU clock rate. 

Similarly I have not benchmarked whether the python overhead of keeping a count so we could call this every x calls would be greater than the overhead of switching to the main thread finding there are no events to process and switching back.

If we did it once a directory instead of once per file, then it would depend on how many files there are in the directory - which or some releases on MB is several hundred.

And remember it is better to do it more often than necessary to keep the UI responsive rather than less often (which would create unresponsiveness).